internal/storage/multi_store.go

```diff
@@ -0,0 +1,445 @@
+	keys := []string{defaultStorage}
+	for k := range s.store {
+		if k == defaultStorage {
+			continue
+		}
+		keys = append(keys, k)
+	}
+	return iteratorMultiStoreServer{
+		multiStoreServer: s,
+		keys:             keys,
+	}
+}
+
+func (i *iteratorMultiStoreServer) initialize() {
+	i.counter = 0
+}
+
+func (i *iteratorMultiStoreServer) next() ImageServer {
+	if i.counter > len(i.multiStoreServer.store)-1 {
+		return nil
+	}
+	if e, ok := i.multiStoreServer.store[i.keys[i.counter]]; ok {
+		i.counter++
+		return e
+	}
+	return nil
+}
+
+type MultiStoreServer interface {
+	GetDefaultStorage() cstorage.Store
+	GetAllStores() []cstorage.Store
+	GetImageServer(driver string) (ImageServer, error)
+	GetStoreForContainer(idOrName string) (cstorage.Store, error)
+	GetStoreForImage(imageID string) (cstorage.Store, error)
+	FromContainerDirectory(id, file string) ([]byte, error)
+	ContainerRunDirectory(id string) (string, error)
+	ContainerDirectory(id string) (string, error)
+	Shutdown(force bool) (layers []string, err error)
+	GraphRoot() string
+	GetImageServerForImage(image string) ([]ImageServer, error)
+	GetStore() MultiStore
+	ListAllImages(ctx *ctypes.SystemContext, filter string) ([]ImageResult, error)
+	ResolveNames(systemContext *ctypes.SystemContext, imageName string) ([]string, error)
+	ImageStatus(systemContext *ctypes.SystemContext, filter string) (*ImageResult, error)
+}
+
+// NewMultiStoreServer creates a new instance of the MultiStoreServer.
+func NewMultiStoreServer(store map[string]ImageServer, istore MultiStore) MultiStoreServer {
+	s := &multiStoreServer{
+		store:      store,
+		multiStore: istore,
+	}
+	s.iterator = createMultiStoreServerIterator(s)
+	return s
+}
+
+func (s *multiStoreServer) GetDefaultStorage() cstorage.Store {
+	return s.multiStore.GetDefaultStorage()
+}
+
+// GetImageServer returns the ImageServer for the selected storage driver. If the driver is empty or not found, the default storage driver is returned
+func (s *multiStoreServer) GetImageServer(driver string) (ImageServer, error) {
+	iserver, ok := s.store[driver]
+	if !ok {
+		v, ok := s.store[s.multiStore.GetDefaultStorageDriver()]
+		if !ok || v == nil {
+			return nil, fmt.Errorf("no image server for default storage driver found")
+		}
+		return v, nil
+	}
+	return iserver, nil
+}
+
+// GetStore returns the MultiStore used for configuring the MultiStoreServer.
+func (s *multiStoreServer) GetStore() MultiStore {
+	return s.multiStore
+}
+
+// ListAllImages lists all the images known by the MultiStoreServer.
+func (s *multiStoreServer) ListAllImages(ctx *ctypes.SystemContext, filter string) (imageResults []ImageResult, lastError error) {
+	s.iterator.initialize()
+	for is := s.iterator.next(); is != nil; is = s.iterator.next() {
+		images, err := is.ListImages(ctx, filter)
+		if err != nil {
+			lastError = wrapMultipleErrors(lastError, err)
+			continue
+		}
+		imageResults = append(imageResults, images...)
```

Comment:
do we need to care for duplicates?

Reply:
@giuseppe it depends. Ilsted this issue under the special notes. There isn't a way how I could print the storage where the image is located. If I print only once then we don't get the information that we actually have 2 (or more) copies of the image. This is still an aspect I don't know how to solve in a nice way